고양이 장난감가게 유효성 추가하기, 회원 관리 만들기

alex-gibeom opened the pull request:
안녕하세요 리뷰어님 :)
이번주도 잘부탁드립니다 😊

이번 주는 저번 주에 학습했던 헥사고날 아키텍쳐를 적용하여 개발해 나가볼 예정입니다!

alex-gibeom committed:
refactor: 기존 베이스코드 구조 변경 - 헥사고날 아키텍쳐 적용

alex-gibeom committed:
feat: 공통 예외 구축

alex-gibeom committed:
feat: mapstruct 의존성 추가 - 객체 간 매핑을 도와주는 라이브러리

alex-gibeom committed:
merge: [Day-1] 기본 베이스 코드 구조 변경, 공통 예외 구축, mapstruct 의존성 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/exceptions/CommonException.java:
```diff
@@ -0,0 +1,26 @@
+ * 자식 Exception으로부터 넘겨받은 status와 message를 RuntimeException으로 전달합니다.
+ */
+@Getter
+public class CommonException extends RuntimeException {
```
CommonException 이라는 이름이 예외의 범위를 광범위하게 잡고 있습니다. 예외가 커버하는 범위를 최소화해보도록 하죠. 일단 이 클래스는 HttpStatus를 갖고 있으니까 HttpException 같은 이름으로 변경하는 것도 고려해 보세요.

alex-gibeom replied:
혹시 `HttpBusinessException`은 어떤가요?
해당 네이밍을 생각한 이유는 모든 비지니스 로직에서 논리적 에러가 발생했을 때 던지는 예외들의 공통 부모이기 때문입니다.
또한 말씀해주신대로 HttpStatus도 포함하고 있기에 `HttpBusiness`라고 생각해보았습니다.

아니면 이렇게 하나의 객체로 모든 비지니스 로직을 묶으려는 의도가 위험할 수도 있는 것일까요?

johngrib replied:
별 거 아니고 결국엔 디버깅을 쉽게 하기 위해서에요. 모든 예외를 하나로 통합해 잡아버리면 코딩할 땐 편하지만 메시지를 정교하게 나눠줘야 하기 때문에 운영을 할 때에는 골치아파지죠. 그러나 제가 권장하고 있을 뿐이죠. 이렇게 해야 한다는 법은 없습니다. 하나로 만드는 건 자유입니다. 납득하기 어렵다면 그대로 하나로 계속 만들어도 괜찮아요. 이 방법이 더 마음에 든다면 저도 이것에 대해 리뷰를 더 안 남기면 되죠. 하지만 프로젝트와 언어와 프레임워크를 꿰뚫는 원칙을 잘 고민해 볼 필요가 있죠. 예를 들어 이런 겁니다. "그냥 전부 HashMap을 써도 되는데 왜 Entity와 DTO를 만드는가?" 같은 거죠. Java에서는 섬세하게 나눠진 타입을 만들어 문제로 접근해갑니다. 타입이 문제 해결의 핵심 도구 중 하나가 되는 거죠. 물론 이게 "답"은 아닙니다. 전혀 다른 접근을 선택한 라이브러리나 언어 ,프레임워크도 있죠. 예를 들어 HashMap으로 모든 것을 처리하는 것을 더 선호하는 언어도 있어요. js 라던가 clojure라던가 php가 그렇죠. 더 Java 다운 코드를 작성하고 싶다면 사고 방식의 핵심에 타입을 두고 논리를 펼쳐나가거나 문제에 접근해보는 것도 훈련해 볼 필요가 있어요. 우리가 만들고 있는 교육용 프로젝트가 실제 프로덕트를 만드는 것이 아니라는 것도 고려해야 하죠. 우리는 연습과 훈련을 하고 있는 중이거든요. 곰곰이 생각해 보시고 코멘트를 달아보세요. 원하시는 방향으로 갈 수 있도록 도와드릴게요.

alex-gibeom replied:
일단 먼저 감사드립니다. 덕분에 다시 한번 Java 언어의 특징을 다시 상기시키면서 성찰도 좀 하게 됐네요. 최대한 객체지향적으로 생각을 접근하려고 노력하고 있는데 더 많이 연습을 해야될 것 같습니다.
Java는 타입이 있는 언어이고 Class를 통해 사용자 정의 타입을 만들 수 있으며, 타입을 적극 활용하는 것이 Java를 잘 사용하는 것인 것 같습니다.

`HttpBusinessException`을 생각했었던 이유는 “모든 비지니스 로직에서의 논리적 예외라는 공통점” 이라는 생각으로 하나로 묶으면 좋을 것 같다고 생각했었는데, 다시 생각해보니까 말씀해주신대로 DTO, Entity를 모두 HashMap으로 사용하는 격인 생각이였던 것 같네요.
모든 비지니스 로직에서 발생하는 논리적 예외일지라도, 각 예외의 주제 카테고리에 맞게 예외 Class를 적절하게 쪼개주는 것도 중요할 것 같습니다. 너무 하나의 객체로만 응집하려는 생각에 사로잡혀있었던 것 같습니다.
다른 언어와 비교해주시면서 설명해주셔서 더 이해가 잘된 것 같습니다. 피드백 너무 감사드립니다 :)

해당 클래스는 말씀해주신대로 `HttpException`으로 변경하는 것이 좋아보이네요.
한가지 궁금한 점은 아래의 **[예시 1]** 처럼 HttpException 관련 하위 예외들을 각 예외 카테고리별로 또 한번 부모로 묶어주고 가지치기로 확장해나가는 것이 좋을지, **[예시 2]** 처럼 공통점이 없다면(ex. httpStatus 코드가 서로 다른 경우) 굳이 하위 카테고리는 안묶어줘도 좋을지 헷갈리네요 😂
어떤식으로 예외 클래스들을 관리하는 것이 좋을지 궁금합니다.

#### [예시 1]
![image](https://user-images.githubusercontent.com/59248326/200741476-00807386-aacd-4bf9-a02b-3557cdf0e961.png)

#### [예시 2]
![image](https://user-images.githubusercontent.com/59248326/200741494-7bd9c71d-46d2-4d15-ae25-9b2d348879e6.png)

johngrib replied:
일단 짚어봐야 할 점이 몇 가지가 있습니다. 일단 예시1과 예시2는 특정 HttpStatus를 모두 예외 클래스로 처리하려 하고 있습니다. 저는 컨트롤러가 Http 침투를 차단하는 역할을 담당하도록 가이드를 드리려 하고 있는데요, 그렇다면 이렇게까지 많은 예외를 굳이 정의할 필요는 없을 겁니다. 컨트롤러가 2개인데 예외 클래스가 수십개이고 이런 건 과하죠. 그리고 특정 Http 상태 코드를 예외로 처리하는 것이 바람직한지도 생각해봐야 합니다. Http 상태 코드는 일종의 리턴값이에요. 애플리케이션이 전달받은 HTTP 요청에 대한 응답으로 돌려주는 거죠. 즉 프로토콜(약속)이므로 응답에 상태 코드를 달아서 보내주는 거고, 도메인에서 HTTP를 굳이 의식할 필요가 없습니다. 지금 작업할 건 Http의 각 상태 코드를 예외 클래스로 코딩하는 건 아니에요. 꽤나 지루한 노가다가 될 거에요. 그리고 이 노가다는 만드는 사람도 재미없을 뿐 아니라 우리가 만들고 있는 서버 애플리케이션의 본질이 아닐 수도 있습니다.

지금 정의할 건 고양이 장난감 가게에서 발생 가능한 예외적으로 처리할 일들에 대한 문제를 정의하는 겁니다. 지금은 HTTP를 담당하는 컨트롤러가 주인공이라고 (작업의 핵심이라고) 생각하고 있어서 Http 예외를 만드는 게 중요하다고 생각하고 있는 걸 수도 있어요. 하지만 HTTP 컨트롤러가 아니라 전혀 다른 종류의 컨트롤러로 대체되는 상황을 생각해 봅시다. 고양이 장난감 가게 사장님이 인터넷과 HTTP를 사용하지 않고 그냥 오프라인 가게 관리용 포스기 전용 프로그램을 만들자고 의사결정을 한다고 생각해 보는 거에요. 그렇다면 HTTP 예외들은 어떻게 될까요? 다 지워야 할까요? HTTP가 입출력용 껍데기라는 것을 염두에 두어야 합니다.

지금 @giibeom 님 머릿속에 코딩이 너무 가득한 것 같아요. 코드는 잠시 미뤄두고 종이를 펴고 장난감 가게에서 어떤 일들이 예외적인 상황인지부터 목록으로 만들어 보세요. 상상력을 최대한 발휘하되, 실제 가게를 생각해 보면서 실용성있는 문제 상황을 가정할수록 우리가 만들고자 하는 목적에 부합하는 결과에 접근하게 될 겁니다. 예를 들어 상품의 가격이 0보다 작은 상황이라던가... 상품의 아이디가 중복된 상황이라던가...

alex-gibeom replied:
와.. 생각해보니 HTTP 상태 코드는 프로토콜의 응답일 뿐인데, 이걸 도메인 쪽에서 담당하도록 하려는 생각만 하고 있었던 것 같네요. 컨트롤러가 외부 환경과 유스케이스를 맞춰주는 어댑터 역할을 하므로, 외부 환경과 관련된 입/출력 관련 작업은 어댑터 쪽에서 하는 것이 유스케이스를 외부로부터 독립적으로 지킬 수 있는 방법인데 이러한 관점을 놓치고 있었던 것 같습니다. 해당 관점을 지키려고 Service에서 Controller로 Entity를 그대로 반환해서, Controller에서 응답 포맷을 변환해주었는데, 모순적으로 HTTP 상태 코드까지는 이 관점이 미처 도달하지 못했네요.
정말 어디서도 쉽게 배울 수 없는 것들을 코드숨에서 많이 배워나가고 있는 것 같습니다😊😊 덕분에 관점이 넓어지는 훈련을 하고 있는 것 같습니다. 감사합니다.

멘토님이 권유해주신대로 코딩을 잠시 내려놓고, 실제 장난감 가게 도메인의 예외 상황에 대해서 곰곰히 생각해보며 정리하였습니다. 컬럼이 그렇게 많지 않아서 다양한 경우의 수가 잘 떠오르지 않는 것 같네요.

#### [장난감 가게에서 발생 가능한 예외 케이스]
1. 상품의 이름과, 메이커와, 가격이 없거나 공백인 상황 (필수)
2. 상품 가격이 0보다 작은 상황
3. 상품의 ID가 중복되는 상황
4. 상품이 없는 상황 (상품 상세조회, 수정, 삭제 시)
    
**추가 상상의 나래**
1. 상품 수량이 있다면 상품 수량이 0보다 작은 상황
2. 가게 잔고가 있다면 고객이 상품 구매 시, 거스름돈을 줄 잔돈이 없는 상황

johngrib replied:
좋습니다. 이 예외를 토대로 작업을 차근차근 진행해 보세요.

ㅎㅎ 이번주 작업은 리뷰하는 입장에서 꽤 즐겁고 재미있네요. @giibeom 님이 그만큼 질문을 잘 해주셔서 그런 것 같기도 합니다. 장난감 가게 예외 케이스가 좀 적긴 하지만 이런 일들부터 어떻게 처리할지 전략을 세워보고, 좀 심심하면 새로운 기능도 추가해가며 진행해보도록 합시다. 학습도 학습이지만 재미도 있겠죠!

alex-gibeom replied:
덕분에 객체지향적인 관점에 한발자국 다가간 것 같습니다. 너무 새롭네요. 실제로 경주마처럼 개발만 하다가, 생각을 정리하고 하니 한층 더 개발할 것들이 눈에 잘 보이는것 같습니다😊

(개발: 74cff0b, 테스트: ed113f3)  해당 기능에서 Validation에 관한 기능을 개발해보았습니다.
추후에 따로 시간 내서 수량과 잔고 같은 추가 기능도 개발해보려고 합니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/exceptions/product/ProductNotFoundException.java:
```diff
@@ -0,0 +1,19 @@
+
+import com.codesoom.assignment.exceptions.CommonException;
+import org.springframework.http.HttpStatus;
+
+public class ProductNotFoundException extends CommonException {
+    private static final String MESSAGE = "상품이 존재하지 않습니다.";
+
+    /**
+     * 상품을 찾지 못했을 때 던집니다.
+     */
+    public ProductNotFoundException() {
```
이 설명은 클래스에 달아주는 것이 좋겠습니다.
```suggestion
/**
 * 상품을 찾지 못했을 때 던집니다.
 */
public class ProductNotFoundException extends CommonException {
    private static final String MESSAGE = "상품이 존재하지 않습니다.";

    public ProductNotFoundException() {
```

alex-gibeom replied:
그러는게 좋을 것 같네요 :)
(dda7473) 수정하였습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/product/adapter/in/web/ProductController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @GetMapping
+    public List<Product> list() {
+        return productUseCase.getProducts();
+    }
+
+    @GetMapping("{id}")
+    public Product detail(@PathVariable final Long id) {
+        return productUseCase.getProduct(id);
+    }
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public Product create(@RequestBody @Valid final ProductRequest productRequest) {
+        return productUseCase.createProduct(productRequest);
+    }
+
+    @PatchMapping("{id}")
+    public Product update(
```
Entity가 Controller를 통해 외부로 유출되지 않도록 래퍼나 DTO를 사용해 보세요. Controller가 애플리케이션과 HTTP가 맞닿는 경계선이라고 생각해봅시다.

alex-gibeom replied:
(eff5c53) (ce2dd61) (77f52f5) 각 계층 간 데이터 전달 객체를 아래와 같이 모두 분리해보았습니다.
- Client -> Controller : RequestDto
- Controller -> Service : Command
- Service -> UseCase(비지니스 로직) : Entity
- Controller -> Client : ResponseDto

추가로 Service에서 Entity를 그대로 반환하여 Controller에서 변환 과정을 거친 이유를 정리해보았습니다.
1. 웹 어댑터(Controller)는 외부 통신 환경과 유스 케이스를 연결해주는 "어댑터" 역할을 해주기 때문에 Controller에서 외부 환경에 맞게 변환과정을 거치는 것이 바람직해 보였습니다.
2. UseCase는 웹 어댑터의 영역을 몰라도 되는 계층이라고 생각하였습니다.
만약 UseCase에서 Entity -> ResponseDto 변환 작업을 하게 되면 웹 어댑터 코드에 종속되어 재사용할 수 없게 되버린다고 생각하였습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/product/application/port/in/ProductUseCase.java:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @return 상품 목록 리턴
+     */
+    List<Product> getProducts();
```
상품의 수가 엄청나게 많아진다면 이 메소드 시그니처로는 곤란할 수도 있겠습니다. 어떤 방법을 고려할 수 있을까요?

alex-gibeom replied:
상품이 100만건, 1000만건 이상이라면 페이징 처리를 통해 일부 내용만 반환하도록 하는 방법이 있을 것 같습니다.

johngrib replied:
역시 그렇겠죠. 여유가 있다면 추가 과제로 페이징 처리를 "구현"해보세요.

giibeom replied:
(cba7b8f) JPA 자체가 처음이다보니까 테스트에서 FakeObject 내에서의 페이징 구현이 좀 어려운 것 같습니다 😂 테스트가 깨지는 것들을 쉽게 잡기가 어려운 것 같습니다.
Pageable 객체를 사용하는 것이 쉽지 않네요..!

johngrib commented on app/src/test/java/com/codesoom/assignment/product/adapter/in/web/ProductControllerTest.java:
```diff
@@ -72,19 +75,19 @@ void setUp() {
-                        .accept(MediaType.APPLICATION_JSON_UTF8)
-        )
+                        get("/products")
+                                .accept(MediaType.APPLICATION_JSON_UTF8)
```
`MediaType.APPLICATION_JSON_UTF8`의 JavaDoc을 읽어보세요. 어떤 내용이 담겨 있나요?

alex-gibeom replied:
`@Deprecated` 된 이유는 Chrome과 같은 주요 브라우저가 charset=UTF-8 매개변수 없이도 UTF-8 특수 문자를 올바르게 해석하므로 `MediaType.APPLICATION_JSON_UTF8`가 아닌 `MediaType.APPLICATION_JSON`을 선호한다고 나와있습니다.

해당 어노테이션을 개발할 때 확인했는데 사용했던 이유는 브라우저를 통한 통신이 아닌, Controller Test(웹 통합 테스트)에서 해당 설정을 안해줄 경우 인코딩이 깨지는 현상이 발생하여 `.andExpect(content().string(containsString(TOY_1.NAME())))` 구문이 전부 테스트 검증 실패로 되어, `@Deprecated`여도 사용을 하였습니다.

혹시 다른 더 좋은 방법이 있을까요? 😂

johngrib replied:
그 방법을 직접 조사해서 기록으로 남겨보세요. 노션이나 블로그 같은 공간에 문서를 만들고 제일 위에 문제 상황을 명시하고, 아래쪽에 문제를 해결해가는 과정을 순차적으로 적는 겁니다. 구글링을 해도 좋고 스택오버플로우를 찾아봐도 좋습니다. 그러나 빠뜨리면 안되는 가장 중요한 건, 문서 하단에 출처를 적는 거에요. 꼭 해보세요. 앞으로도 다양한 문제를 만나게 될 텐데 이렇게 기록으로 남기면서 작업해보면 좋은 훈련이 될 것입니다.

alex-gibeom replied:
멘토님이 권유해주신 대로 [블로그](https://velog.io/@beomdrive/MockMvc-%ED%85%8C%EC%8A%A4%ED%8A%B8-%EC%8B%9C-%EC%9D%B8%EC%BD%94%EB%94%A9-%EB%AC%B8%EC%A0%9C%EB%A1%9C-%EA%B2%80%EC%A6%9D-%EC%8B%A4%ED%8C%A8)를 통해 해당 문제를 상황, 원인, 해결, 참고 자료 순으로 차근차근 정리해서 작성해보았습니다.
블로그에 익숙하지 않아 오늘 하루 종일 투자하였는데도, 성에 차진 않은 것 같네요 😂

요약하자면 Chrome 같은 메인 브라우저들이 charset이 없어도 UTF-8 특수문자를 잘 해석하기 때문에 스프링에서는 charset을 사용하지 않는 방향으로 나아갔고, 그로 인해 `MediaType.APPLICATION_JSON_UTF8` 또한 자연스럽게 Deprecated가 된 것 같습니다.
따라서 스프링의 본래 의도에 벗어나지 않도록 MockMvc의 테스트 결과값에 `CharacterEncoding`을 UTF-8로 설정되도록 수정하였습니다.

johngrib replied:
아 좋습니다. 너무x100 잘하셨어요. 우리가 이렇게 코멘트로 대화하는 것보다 이렇게 혼자 생각하고 실험하며 완성된 글로 남기는 경험이 훨씬 중요한 경험일 수 있습니다. 제가 좋아하는 방식이기도 한데요, 자기 자신을 과학자라고 생각하고 문제에 접근하는 것입니다. 대상을 정의하고, 실험을 설계하고, 실험을 실행해 본 다음 결과까지 기록하고, 그 결과를 누구나 볼 수 있는 곳에 남겨두는 거죠. 이런 기록이 많아져서 누가 보고 채용 기회가 생기고 어쩌고 하는 것은 별로 중요한 일이 아니라고 생각합니다. 스스로 과학자라고 생각하고 이렇게 접근하고 실험하고 하는 일이 무엇보다 @giibeom 님 자신의 사고 방식과 일하는 방식을 더 정교하고 깊이있게 만들어 줄 거에요.

alex-gibeom replied:
생각의 흐름을 블로그에 정리하는 것이 익숙하지 않아 시간이 너무 오래걸렸지만, 오히려 깊이 있는 학습이 된 것 같아서 좋은 경험을 하게 된 것 같습니다. 새로운 사실도 알게 되었구요 😊
앞으로도 계속해서 `대상을 정의하고, 실험을 설계하고, 실험을 실행해 본 다음 결과까지 기록`하는 습관을 꾸준히 들여야겠습니다. 감사합니다!

johngrib commented on app/src/test/java/com/codesoom/assignment/product/adapter/in/web/ProductControllerTest.java:
```diff
@@ -98,82 +101,82 @@ void deatilWithNotExsitedProduct() throws Exception {
+                        patch("/products/1")
+                                .accept(MediaType.APPLICATION_JSON_UTF8)
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content("{\"name\":\"쥐순이\",\"maker\":\"냥이월드\"," +
+                                        "\"price\":5000}")
```
content의 내용을 문자열로 작성하고 있으니 아쉽습니다. 객체를 통해 접근하는 방법을 고려해 봅시다.

alex-gibeom replied:
앗 5주차 기본 베이스 코드가 리뷰 주신 내용대로 되어있었나봅니다.
기본 베이스 코드에서 사용된 모든 mock 기반 테스트를 아래와 같이 FakeObject를 통해 실제 객체로 테스트할 수 있도록 변경해보았습니다.
또한 Fixture 데이터도 따로 enum객체를 사용하여 재사용할 수 있도록 해보았습니다.

(73ed70c) JPA 테스트
(1cca503) DB FakeObject 생성 및 테스트
(1b51e19) Service에 있었던 기존 mock 테스트 전부 제거 및 FakeObject를 통해 실제 객체로 테스트하도록 변경
(7b73a37) Controller에 있었던 기존 mock 테스트 전부 제거 및 FakeObject를 통해 실제 객체로 테스트하도록 변경

johngrib reviewed (commented):
이번주도 화이팅!

alex-gibeom committed:
refactor: 공통 예외명 변경 및 주석 보완

alex-gibeom committed:
feat: 계층 간 DTO 생성 (Request, Command, Entity, Response) 및 각 객체 간 converter 생성

alex-gibeom committed:
refactor: 상품 생성할 때 Controller에서 Service로 데이터 전달 시 Command로 변환하여 전달하도록 변경 (RequestDto -> Command)

alex-gibeom committed:
feat: ProductPersistenceAdapter JPA 단위 테스트 추가

alex-gibeom committed:
feat: ProductPersistenceAdapter FakeObject 생성 및 단위 테스트 추가 (InMemory Data)

alex-gibeom committed:
refactor: ProductService 단위 테스트 - 모든 mock 테스트 제거, FakeObject를 사용하여 실제 객체로 테스트 수행하도록 일괄 변경

- 기존 5주차 베이스 코드에 있는 mock 테스트를 전부 삭제 후 실제 객체로 테스트 수행

alex-gibeom committed:
feat: Jackson 라이브러리를 사용한 JsonUtil 추가

alex-gibeom committed:
refactor: ProductController 웹 통합 테스트 - 모든 mock 테스트 제거, FakeObject를 사용하여 실제 객체로 테스트 수행하도록 일괄 변경

- 기존 5주차 베이스 코드에 있는 mock 테스트를 전부 삭제 후 실제 객체로 테스트 수행

alex-gibeom committed:
refactor: 상품 생성 응답 데이터를 Entity에서 ResponseDto로 변환하여 응답하도록 변경 (Entity -> ResponseDto)

alex-gibeom committed:
feat: Entity 리스트에서 ResponseDto 리스트로 매핑하는 converter 추가

alex-gibeom committed:
refactor: Controller 응답 데이터를 Entity에서 ResponseDto로 변환하여 응답하도록 변경 (Entity -> ResponseDto)

alex-gibeom committed:
merge: [Day-2] 기본 베이스 테스트 코드 모두 변경, 계층 간 데이터 전달 객체 분리

- 계층 간 데이터 전달 객체
   - Client -> Controller : RequestDto
   - Controller -> Service : Command
   - Service -> 유스케이스 : Entity
   - Controller -> Client : ResponseDto

- 객체 간 매핑 converter 구현 (MapStruct)

- 모든 mock 기반 테스트 전부 mock 제거 작업
   - FakeObject를 통해 실제 객체로 테스트 진행

alex-gibeom commented on app/src/main/java/com/codesoom/assignment/product/application/port/in/ProductMapper.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.util.List;
+
+@Mapper(componentModel = "spring")
+public interface ProductMapper {
+    ProductMapper INSTANCE = Mappers.getMapper(ProductMapper.class);
+
+    /**
+     * RequestDto 객체에서 Command 객체로 매핑합니다.
+     */
+    ProductCommand requestToCommand(ProductRequest productRequest);
+
+    /**
+     * Command 객체에서 Entity 객체로 매핑합니다.
+     */
+    @Mapping(target = "id", ignore = true)
+    Product commandToEntity(ProductCommand productCommand);
+
+    /**
+     * Entity 객체에서 ResponseDto 객체로 매핑합니다.
+     */
+    ProductResponse entityToResponse(Product entity);
+
+    /**
+     * Entity 객체의 리스트에서 ResponseDto 객체의 리스트로 매핑합니다.
+     */
+    List<ProductResponse> entityListToResponseList(List<Product> entity);
```
5주차 강의에서는 Dozer Mapper 라이브러리를 통해 객체 간 매핑을 해주었습니다.
하지만 서칭해본 결과, Dozer Mapper 라이브러리는 내부적으로 리플렉션을 사용하여 성능이 좋지 않다고 합니다.
따라서 어노테이션 프로세서 기반으로 동작하는 MapStruct를 사용하여 객체 간 매핑 기능을 구현해보았습니다!

여기서 한가지 질문이 있습니다.
모든 도메인에서 동일하게 RequestDto -> Command -> Entity -> ResponseDto 변환이 진행할 것으로 보이는데, 해당 인터페이스를 공통 객체로 추상화를 하고싶은데, 어떤식으로 진행해야될 지 모르겠습니다...😂
제네릭을 사용해서 구현해보려고 했는데 실패하였습니다...
조금만 가이드를 주실 수 있으신지요?

johngrib replied:
오 트레이드 오프를 고려해 Dozer Mapper를 탈락시킨 결정을 하셨네요. 아주 잘 하셨습니다. 그런데 이렇게 코멘트로만 남기지 마시고 문서를 만들어서 다른 라이브러리와 성능 비교를 해보았다면 더 좋았을 겁니다. "좋지 않다고 합니다" 라는 표현이 무척 아쉬워요.

한편 Dozer Mapper는 다른 문제가 하나 더 있기 때문에 Dozer Mapper로 작업을 하셨다면 제가 Dozer Mapper를 제거해보라는 추가 과제를 드렸을 겁니다. (이 문제가 어떤 것인지는 Dozer Mapper github에 가보시면 금방 알 수 있을 거에요.) 하지만 여기에서는 아예 Dozer Mapper를 처음부터 탈락시켰으니 다른 추가 과제를 드리도록 할게요. MapStruct의 침투율과 제거 시나리오에 대한 1~2 페이지짜리 보고서를 작성해 보세요. 회사에서 동료들을 설득하기 위한 진지한 자료를 만든다고 생각하시면 됩니다. 그리고 MapStruct를 제거해야 하는 상황을 대비해 어떤 방식으로 MapStruct를 사용할지에 대해서 계획이 들어 있으면 됩니다. (실제로 MapStruct를 제거하는 코딩을 하라는 것은 아닙니다. 코드는 그대로 두시고 글을 쓰시면 됩니다.)

RequestDto -> Command -> Entity -> ResponseDto 변환에 대해 고민하고 계시군요. 보통은 불편하다고 생각은 해도 질문을 안 하고 넘어가는데 솔직히 질문을 해 주시니 다행입니다. 사실 이건 Java를 쓰는 이상 빠져나가기 힘든 일종의 트랩이기도 합니다. 내부 구현 유출에 대한 감각이 없는 경우엔 Entity를 그냥 HTTP 응답으로 내보내는 일도 흔하게 일어나죠. 지금 작업하고 있는 교육용 프로젝트는 구조가 꽤 단순해서 DTO랑 Entity가 거의 구분이 가질 않고 그게 그거로 보여서 굉장히 답답할 거에요. 여기엔 다양한 해법이 있는데, 제 노하우를 하나 공유하는 것도 의미가 있을 것 같습니다. 물론 이 방법은 수많은 방법 중의 하나입니다.

바깥에서 안으로 들어가는 방향 먼저 보죠.

- Command 는 interface로 정의합니다. 이 인터페이스는 몇 가지 getter 를 갖습니다.
- ReqeustDTO가 Command interface를 구현하게 합니다.
- service의 몇몇 메소드는 입력 파라미터로 DTO나 엔티티를 받지 않습니다. Command interface를 입력받습니다.

예제를 들어야 이해가 더 쉽겠네요.

- ProductCreateRequest 라는 interface를 만듭니다. 이 interface는 getName, getPrice 같은 상품 생성에 필요한 값을 제공하는 메소드들만 정의하고 있습니다. 여기에서는 생각하기 쉽게 이 두 가지만 있다고 칩시다.
- ProductService::create 메소드는 ProductCreateRequest 인터페이스 타입의 인자 하나만 받습니다. 즉 create 는 name 과 price 만 사용하게 됩니다. 그러나 어떤 값이 필요한지는 ProductCreateRequest 라는 이름 아래에 가려집니다. 상품 생성 메소드가 상품 생성 요청을 받는다는 상식적인 영어 표현만 남게 됩니다.
- ProductCreateRequestDto 는 ProductCreateRequest interface를 구현합니다. 따라서 controller는 DTO를 입력받은 상태 그대로 어떠한 변환도 하지 않고 productService::create 메소드에 넘겨주기만 하면 됩니다.
- 이 다음 작업들도 제가 일일이 이야기하는 건 아쉬운 일이겠죠. 이 다음은 상상해서 메꿔 보세요.

안에서 바깥으로 나갈 때에는 저는 상황에 따라 이렇게 하곤 합니다.

- 컨트롤러 안쪽까지는 그냥 Entity를 씁니다. 여기까지는 Entity의 통제범위라고 생각하고 보통 작업하곤 합니다. 저는 헥사고날을 선호하지 않아서 그냥 Entity가 애플리케이션 레이어 안에서 돌아다니게 합니다.
- 컨트롤러 바깥으로 내보낼 때에는 Entity의 wrapper가 되도록 DTO 클래스를 설계합니다. DTO 클래스의 생성자는 Entity 하나를 받는데, DTO has a Entity 개념으로 만듭니다. 단, `@JsonIgnore` 같은 걸 달아서 생성된 json 이 entity의 모든 필드를 유출하지 않게 합니다. DTO는 Entity 값을 감싸서 특별한 형태로 보여주는 역할을 하는 것이라는 개념으로 접근한 거죠. 지금 프로젝트처럼 구조가 단순하면 이 방법이 별로 매력이 없습니다. 하지만 보여줘야 하는 값들이나 포맷에 가공이 필요하다면 일단 기억해두면 언젠간 쓸 일이 있을 거에요.

giibeom replied:
(구현해보고 있습니다..!)

alex-gibeom replied:
> MapStruct의 침투율과 제거 시나리오에 대한 1~2 페이지짜리 보고서를 작성해 보세요. 회사에서 동료들을 설득하기 위한 진지한 자료를 만든다고 생각하시면 됩니다. 그리고 MapStruct를 제거해야 하는 상황을 대비해 어떤 방식으로 MapStruct를 사용할지에 대해서 계획이 들어 있으면 됩니다. (실제로 MapStruct를 제거하는 코딩을 하라는 것은 아닙니다. 코드는 그대로 두시고 글을 쓰시면 됩니다.)

MapStruct 관련 보고서는 약간 오래 걸릴 것 같아서 이번주까지는 **꼭 작성**해보겠습니다. 제거해야 되는 상황에 대한 대비책은 대략적으로 어떻게 작성해야될 지 알 것 같습니다.
한가지 궁금한점은 **침투율**이라는 것이 어떤 의미인지 잘 이해가 되지 않습니다..! 😂

---

> 여기엔 다양한 해법이 있는데, 제 노하우를 하나 공유하는 것도 의미가 있을 것 같습니다. 물론 이 방법은 수많은 방법 중의 하나입니다.
바깥에서 안으로 들어가는 방향 먼저 보죠.
Command 는 interface로 정의합니다. 이 인터페이스는 몇 가지 getter 를 갖습니다.
ReqeustDTO가 Command interface를 구현하게 합니다.
service의 몇몇 메소드는 입력 파라미터로 DTO나 엔티티를 받지 않습니다. Command interface를 입력받습니다.

멘토님의 노하우를 알려주셔서 정말 감사드립니다!
Command를 Class로 구현하기보다는 멘토님 추천대로 Interface로 구현하는 것이 Controller -> Service의 어쩔 수 없었던 의존 방향을 역전시키는 방법인 것 같네요.
(6a9c288) 해당 피드백을 주시자마자 흥미로워서 바로 구현해보았습니다.

---

> 안에서 바깥으로 나갈 때에는 저는 상황에 따라 이렇게 하곤 합니다.
컨트롤러 바깥으로 내보낼 때에는 Entity의 wrapper가 되도록 DTO 클래스를 설계합니다. DTO 클래스의 생성자는 Entity 하나를 받는데, DTO has a Entity 개념으로 만듭니다. 단, `@JsonIgnore` 같은 걸 달아서 생성된 json 이 entity의 모든 필드를 유출하지 않게 합니다. DTO는 Entity 값을 감싸서 특별한 형태로 보여주는 역할을 하는 것이라는 개념으로 접근한 거죠. 지금 프로젝트처럼 구조가 단순하면 이 방법이 별로 매력이 없습니다. 하지만 보여줘야 하는 값들이나 포맷에 가공이 필요하다면 일단 기억해두면 언젠간 쓸 일이 있을 거에요.

해당 부분도 바로 구현해보려고 했는데 아래 리뷰(https://github.com/CodeSoom/spring-week5-assignment-1/pull/80#discussion_r1017875646) 에서 넓은 관점으로 생각할 수 있는 기회를 제공해주셔서 중단하였습니다.
아래 피드백에 대한 생각을 마친 뒤 구현해 볼 예정입니다. 감사합니다!

johngrib replied:
제가 침투율에 대해 @giibeom 님에게 설명하지 않고 막 사용하고 있었네요. 공식적인 의미가 있거나 한 건 아니고 특정 방향에서 스며들듯 침투해 들어가는 기술을 설명할 때 제가 자의적으로 쓰는 단어입니다. 거의 모든 기술에 이 단어를 갖다붙일 수가 있어서 주의할 필요가 있어요. 사용 예는 그리 어렵진 않은데 대충 이렇게 쓸 수 있습니다.

- 프로젝트 전체에 롬복 애노테이션을 쓴 상황: "롬복 침투율이 상당해서 롬복을 쉽게 들어낼 수 없겠어요."
- 프론트엔드에서 SQL을 만들어 서버로 전달하는 괴상하고 위험한 프로젝트: "SQL이 역으로 프론트엔드까지 침투해 들어갔네요?"

alex-gibeom committed:
refactor: MockMvc 웹 통합 테스트 결과값의 CharacterEncoding UTF-8로 설정

- @AutoconfigureMockMvc를 사용하여 MockMvc를 주입하는 경우에 한해서만 적용됨

alex-gibeom committed:
merge: [Day-3] MockMvc 웹 통합 테스트 결과값의 CharacterEncoding UTF-8로 설정

johngrib commented on app/src/main/java/com/codesoom/assignment/exceptions/product/ProductNotFoundException.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * 상품을 찾지 못했을 때 던지는 예외입니다.
+ */
+public class ProductNotFoundException extends HttpBusinessException {
+    private static final String MESSAGE = "상품이 존재하지 않습니다.";
+
+    public ProductNotFoundException() {
+        super(MESSAGE, HttpStatus.NOT_FOUND);
+    }
+
```
"찾을 수 없는 상품 예외"가 HTTP에 의존하고 있습니다. HTTP가 특수한 출력 방식이라는 것을 떠올려보면서 이 의존을 분리해 보세요. 지금 만들고 있는 애플리케이션이 다양한 출력을 지원해야 할 수 있음은 물론, 출력 자체를 알지 못하도록 만드는 방향으로 생각하며 작업해 보세요.

물론 우리가 실제로 회사에서 일하면서 가장 많이 하는 일은 HTTP를 다루는 것이기 때문에 이렇게 예외 클래스를 디자인하는 것은 실용적인 접근일 수도 있어요. 따라서 분리해보라는 제안이 이상하게 느껴질 수도 있습니다. 하지만 @giibeom 님은 과제가 아닌데도 스스로 헥사고날을 연습하겠다고 하셨어요. 따라서 이런 종류의 의존을 떼어내는 연습을 하는 것이 @giibeom 님께는 의미있는 연습이 될 거라 생각합니다. 물론 떼어내는 작업만 하면 안되고 열심히 생각도 해보셔야 해요. 왜 떼어내야 하는가? 객체지향에서 의존이란 무엇인가? 나아가 코딩의 영역을 넘어 개념상의 의존에 대해서도 생각해 보시기를 권해드리고 싶습니다.

alex-gibeom replied:
멘토님이 권유해주신대로 [블로그](https://velog.io/@beomdrive/Exception%EA%B3%BC-HTTP%EB%A5%BC-%EB%B6%84%EB%A6%AC%ED%95%B4%EB%B3%B4%EC%9E%90)를 통해 예외 객체에서 HTTP를 왜 떼어내야 하는지, 객체지향에서의 의존은 무엇인지, HTTP 분리 작업 과정등을  정리해보았습니다😊

(a6c0941, 7f38d96) 작업 내용은 기존 HttpBusinessException을 제거하고, 예외 클래스에서 설정하던 HTTP Status를 `@RestControllerAdvice` 내부 `@ExceptionHandler` 메서드에서 설정하도록 변경하였습니다. https://github.com/CodeSoom/spring-week5-assignment-1/pull/80#discussion_r1018679697 해당 리뷰에 대한 피드백을 받은 후에 Exception 객체들의 관리에 대한 추가 작업을 해볼 예정입니다.

현재 변경한 Exception 객체의 구조는 현재 아래와 같은 형태입니다.
(`ProductNotFoundException`을 제외한 나머지는 예시입니다!)
<img src="https://user-images.githubusercontent.com/59248326/201071890-ae884fb4-2b66-4a09-978f-cafcb8d547f6.png" width="80%" />

alex-gibeom replied:
해당 부분을 나중에 도메인이 더 추가되는 것을 대비하여 `NotFoundException`과 `InvalidException`으로 각각 묶어주는 방법은 어떨지 궁금합니다!

johngrib replied:
그렇게 묶는 것이 도메인 특성상 의미가 있다면 묶어 보도록 해보세요. 그런데 NotFoundException 의 경우 나중에 애매한 위치가 될 가능성이 있습니다. 예를 들어 '고객을 못 찾는 경우', '상품을 못 찾는 경우', '쿠폰을 못 찾는 경우', '주문서를 못 찾는 경우', '카테고리를 못 찾는 경우'는 모두 무언가를 못 찾는 경우이긴 하지만 하나로 묶어주게 된다면 의미없는 연결고리만 추가해주는 것일 수도 있습니다.

alex-gibeom replied:
아하 그런 상황이 벌어질 수 있겠네요. 경험이 부족하다보니 여러 상황에 대한 예측이 어려운것 같습니다 😂

멘토님의 피드백을 듣고 나니 차라리 Exception을 공통 부모로 직접 묶기보다는, 각각의 카테고리에 대한 `@ControllerAdvice` 클래스를 만들고, 각각의 Exception 별로 핸들링해주는게 예외 카테고리 별 응집도도 높이면서 개별적으로 핸들링해주기도 좋을 것 같다는 생각이 드네요!

```java
@RestControllerAdvice
public class NotFoundExceptionHandler {

//  ProductNotFoundException에 대한 개별 처리
    @ExceptionHandler(value = {ProductNotFoundException.class})
    protected ResponseEntity<ErrorResponse> handleProductNotFoundException(final ProductNotFoundException exception) {
    }

//  UserNotFoundException에 대한 개별 처리
    @ExceptionHandler(value = {UserNotFoundException.class})
    protected ResponseEntity<ErrorResponse> handleUserNotFoundException(final UserNotFoundException exception) {
    }
}
```

```java
@RestControllerAdvice
public class InvalidExceptionHandler {

//  ProductInvalidException에 대한 개별 처리
    @ExceptionHandler(value = {ProductInvalidException.class})
    protected ResponseEntity<ErrorResponse> handleProductInvalidException(final ProductInvalidException exception) {
    }

//  UserInvalidException에 대한 개별 처리
    @ExceptionHandler(value = {UserInvalidException.class})
    protected ResponseEntity<ErrorResponse> handleUserInvalidException(final UserInvalidException exception) {
    }
}
```

johngrib replied:
이런 상황에서는 Product와 관련된 예외들을 하나의 어드바이스로 묶고, User와 관련된 예외들을 하나의 어드바이스로 묶는 것을 추천합니다. 실제 회사에서는 상품팀과 회원팀이 서로 관리하는 코드가 다른 경우가 많아서 보여주신 예시처럼 둘이 섞여 있으면 두 팀 사람들이 일하다가 답답해서 상품별/회원별로 별도로 리팩토링하게 될 가능성이 있습니다.

alex-gibeom replied:
아하..! 그렇겠군요!! 덕분에 예외 객체들의 관리를 어떤식으로 해나가야될지 확실하게 이해한 것 같습니다.
감사합니다 :)

alex-gibeom replied:
(0e63019) 멘토님께서 추천해주신 방향대로 각 도메인별로 `ControllerAdvice`를 생성하여 각각의 예외 처리를 도메인 단위로 관심사를 분리해주었습니다!
기존엔 `@RestControllerAdvice(basePackage="")`를 사용하려다가 패키지 경로를 문자열로 직접 치는 것은 위험할 것으로 판단되어 `basePackageClasses`를 사용해보았습니다. 또한 도메인 관련 컨트롤러가 많아질 것을 대비해 커스텀 어노테이션을 통해 설정하도록 지정해주었습니다!

#### [커스텀 어노테이션]
```java
@Target(ElementType.TYPE)
@Retention(RetentionPolicy.RUNTIME)
@RestControllerAdvice(basePackageClasses = {
        UserController.class
})
@Order(Ordered.HIGHEST_PRECEDENCE)
public @interface UserControllerAdvice {
}
```

johngrib commented on app/src/main/java/com/codesoom/assignment/product/adapter/in/web/ProductController.java:
```diff
@@ -0,0 +1,64 @@
+
+    @PostMapping
+    @ResponseStatus(HttpStatus.CREATED)
+    public ProductResponse create(@RequestBody @Valid final ProductRequest productRequest) {
+        return ProductResponse.from(
+                productUseCase.createProduct(productRequest.toCommand())
+        );
+    }
+
+    @RequestMapping(path = "/{id}", method = {RequestMethod.PUT, RequestMethod.PATCH})
+    public ProductResponse update(@PathVariable final Long id,
+                                  @RequestBody @Valid final ProductRequest productRequest) {
+        return ProductResponse.from(
+                productUseCase.updateProduct(id, productRequest)
+        );
+    }
```
만약 생성 리스폰스와 업데이트 리스폰스의 포맷이 달라야 하는 과제가 주어졌다고 생각해 봅시다. 어떻게 대응할 수 있을까요? (코딩하지 마시고 생각만 해서 답변을 달아 보세요.)

- 생성 리스폰스는 생성된 상품의 아이디와 생성된 시간만 포함해야 합니다.
- 업데이트 리스폰스는 "고양이 낚싯대(색깔: 분홍) 상품 (아이디 3번)이 성공적으로 업데이트 되었습니다."라는 문자열만 갖고 있어야 합니다.

그리고 3일 뒤에 사장님이 생각이 바뀌어서 다음과 같이 요구조건이 또 바뀌었습니다.

- 생성 리스폰스는 생성된 상품의 데이터 필드를 모두 갖고 있습니다. 단 생성 시간은 한국 시간이 아니라 UTC로 제공해야 합니다..
- 업데이트 리스폰스는 "고양이 낚싯대(색깔: 분홍) 상품 (아이디 3번)이 성공적으로 업데이트 되었습니다. 업데이트된 시간은 UTC로 12:00:00 입니다."라는 문자열만 갖고 있어야 합니다.

alex-gibeom replied:
> 만약 생성 리스폰스와 업데이트 리스폰스의 포맷이 달라야 하는 과제가 주어졌다고 생각해 봅시다. 

말씀하신대로 요구 사항에서 생성하기, 수정하기 Response가 다를 경우에는 RequestDto때 처럼 분리해야될 것 같습니다.
- CreateResponseDto
- UpdateResponseDto

---

> 생성 리스폰스는 생성된 상품의 아이디와 생성된 시간만 포함해야 합니다.
업데이트 리스폰스는 "고양이 낚싯대(색깔: 분홍) 상품 (아이디 3번)이 성공적으로 업데이트 되었습니다."라는 문자열만 갖고 있어야 합니다.

https://github.com/CodeSoom/spring-week5-assignment-1/pull/80#discussion_r1016577933 에서 말씀해주신 DTO has Entity 방식을 참고하자면 다음과 같이 진행할 것 같습니다. (Entity를 DTO 멤버에 담아서 감춰(`@JsonIgnore`)놓는 방법)
- CreateResponseDto에서는 아이디와 생성된 시간의 변수만 생성하고, Entity에서 값을 세팅하여 반환할 것 같습니다.
- UpdateResponseDto에서는 문자열을 담는 변수 하나를 선언하고, Entity에서 값을 꺼내와 원하는 문자열로 세팅하여 초기화 한 후 반환할 것 같습니다.

---

> 그리고 3일 뒤에 사장님이 생각이 바뀌어서 다음과 같이 요구조건이 또 바뀌었습니다.
생성 리스폰스는 생성된 상품의 데이터 필드를 모두 갖고 있습니다. 단 생성 시간은 한국 시간이 아니라 UTC로 제공해야 합니다..
업데이트 리스폰스는 "고양이 낚싯대(색깔: 분홍) 상품 (아이디 3번)이 성공적으로 업데이트 되었습니다. 업데이트된 시간은 UTC로 12:00:00 입니다."라는 문자열만 갖고 있어야 합니다.

위와 동일한 방법으로
- CreateResponseDto에서는 모든 필드를 선언하고 Entity에서 값을 가져와 세팅한 후, 생성 시간은 UTC로 변환하여 세팅 후 반환할 것 같습니다.
- UpdateResponseDto에서는 Entity에서 값을 가져와 원하는 문자열로 세팅하고, 마찬가지로 업데이트 시간을 UTC로 변환하여 세팅 후 반환할 것 같습니다.

---

이렇게 응답 반환 값을 각기 다른 복잡한 요구 사항을 생각해보니 ResponseDTO에서 Entity를 가지고 있고, 포맷에 맞게 변환해주는 것이 매우 현명한 방법인 것 같네요!!
Entity를 DTO 안에서 갖고 있지 않다면 해당 요구 사항을 반영하기에는 너무 어려울 것으로 보입니다.
ResponseDTO의 역할이 Entity를 원하는 포맷에 맞게 변환해서 보여주는 객체라는 관점에서 Entity를 안에서 품고 있는 것이 좋을 것 같습니다😊

johngrib replied:
내부에 원본 데이터를 보존하고 있으면 다양한 요청에 대해 각 getter 메소드를 다시 코딩하는 것으로 변화하는 상황에 대응할 수 있겠죠. 사실 이렇게 내부에 원본 데이터를 보존하는 것도 일종의 소규모로 만든 repository라 할 수 있을 것입니다. 즉, DTO는 원본 데이터를 특수하게 비틀어 보여줄 뿐인 view가 되는 거죠.

alex-gibeom replied:
(981ede0) 해당 커밋에서 멘토님이 권유해주신 ResponseDTO를 Dto has a Entity 구조로 변경해보았습니다.

추가로 별다른 변수 없이 getter 메서드만으로 반환된다는 것이 잘 이해가 되지 않아 [블로그를 통해](https://velog.io/@beomdrive/RestController%EC%97%90%EC%84%9C-ResponseDto%EB%A1%9C-%EB%B0%98%ED%99%98%ED%95%98%EA%B8%B0-MappingJackson2HttpMessageConverter) 관련 내용을 상황-분석-검색-테스트 순으로 이해하는 과정을 정리해보았습니다!

RestController에서 반환 시 MessageConverter가 MappingJackson2HttpMessageConverter로 위임한 후 Jackson에서 데이터를 매핑할 때 자바의 프로퍼티(getter)를 사용해서 매핑하는 것으로 보입니다. 따라서 변수가 없이도 getter만으로 데이터가 정상적으로 잘 반환되는 것 같네요.
또 하나의 새로운 사실을 알게된 것 같아 뿌듯한 것 같습니다!

alex-gibeom committed:
refactor: Command 객체(Controller -> Service 데이터 전달 객체) 인터페이스화 진행 및 생성&수정 RequestDTO 분리 작업

alex-gibeom committed:
merge: [Day-4] Command 객체 인터페이스화 및 생성&수정 RequestDTO 분리 작업

alex-gibeom committed:
refactor: Exception 객체에서의 HTTP 의존성 분리 작업

alex-gibeom committed:
merge: [Day-4] Exception 객체에서의 HTTP 의존성 분리 작업

alex-gibeom committed:
fix: ProductNotFoundException 에러 시 반환코드 404 설정 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/product/application/port/in/command/ProductCreateRequest.java:
```diff
@@ -0,0 +1,15 @@
+
+import com.codesoom.assignment.product.domain.Product;
+
+public interface ProductCreateRequest {
+    String getName();
+
+    String getMaker();
+
+    Integer getPrice();
+
+    String getImageUrl();
+
+    Product toEntity();
+}
```
좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/product/application/port/in/command/ProductUpdateRequest.java:
```diff
@@ -0,0 +1,15 @@
+
+import com.codesoom.assignment.product.domain.Product;
+
+public interface ProductUpdateRequest {
```
`ProductCreateRequest`와 별도의 파일로 만든 선택이 좋습니다. `ProductCreateRequest`와 똑같은 추상 메소드 정의를 갖고 있지만 두 인터페이스의 내부 명세가 같은 것은 '우연'이죠. 생성 요청은 생성 요청만의 은닉된 공간으로, 업데이트 요청은 업데이트 요청만의 은닉된 공간을 갖는다는 관점으로 잘 접근하신 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/product/application/ProductService.java:
```diff
@@ -28,19 +28,14 @@ public Product getProduct(Long id) {
     }
 
-    public Product updateProduct(final Long id, final ProductRequest productRequest) {
+    public Product updateProduct(final Long id, final ProductUpdateRequest productUpdateRequest) {
```
좋습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/product/adapter/in/web/dto/ProductUpdateRequestDto.java:
```diff
@@ -0,0 +1,30 @@
+
+    private String imageUrl;
+
+    @Override
+    public Product toEntity() {
+        return ProductMapper.INSTANCE.toEntity(this);
+    }
```
좋습니다. 이런 기능은 따로 묶어 factory 클래스를 만들기도 합니다. 지금은 이대로도 충분해 보입니다.

alex-gibeom committed:
refactor: Product의 ResponseDTO를 Dto has a Entity 구조로 변경 (CreateResponse, UpdateResponse)

alex-gibeom committed:
feat: 상품 Product Validation 기능 추가

- 상품명: null, 공백 허용 안함
- 메이커: null, 공백 허용 안함
- 가격: null, 음수 허용 안함

alex-gibeom committed:
test: 상품 Product Validation 웹 mvc 테스트 추가

- 상품명 : 공백
- 메이커 : 공백
- 가격 : 음수

alex-gibeom committed:
feature: Product 상품 목록 조회 페이징 기능 추가 (진행중)

alex-gibeom committed:
merge: [Day-5] ResponseDTO 작업 완료, Validation 작업 완료, 상품 목록 페이징 기능 개발중

johngrib commented on app/src/main/java/com/codesoom/assignment/product/adapter/in/web/dto/response/CreateProductResponseDto.java:
```diff
@@ -0,0 +1,33 @@
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+public class CreateProductResponseDto {
+    @JsonIgnore
+    private Product product;
```
좋습니다. 이제 이 방식에는 상당히 익숙해지신 것 같군요. 상황에 맞게 유연하게 (이 방식을 쓰기도 하고, 안 쓰기도 하고) 사용하시길 바랍니다. 주로 팀 동료들과 대화해서 결정해 나가면 됩니다.

alex-gibeom replied:
다양한 개발 방법을 알려주셔서 감사합니다. 앞으로 "기술은 정답이 없으니 상황에 따라 유연하게 트레이드 오프를 고려하여 사용하자"라는 생각을 항상 상기시키며 개발하도록 하겠습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/exceptions/GlobalExceptionHandler.java:
```diff
@@ -1,18 +1,29 @@
+    }
+
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(MethodArgumentNotValidException exception, HttpHeaders headers, HttpStatus status, WebRequest request) {
```
한 줄이 너무 길군요. 개행 문자를 중간에 넣어서 두 줄로 나눠도 충분할 것 같습니다.

alex-gibeom replied:
앗 미처 신경쓰지 못한 부분이었네요. 인텔리제이에서 지원해주는 코드 폭의 가이드라인을 잘 확인하겠습니다.
감사합니다 :)

<img src="https://user-images.githubusercontent.com/59248326/201375380-bb6247ec-c87f-4855-91c6-0b9bf0e0fad2.png" width ="80%" />

johngrib commented on app/src/test/java/com/codesoom/assignment/product/adapter/in/web/ProductControllerTest.java:
```diff
@@ -87,26 +90,82 @@ void it_responses_404() throws Exception {
+                    mockMvc.perform(
+                                    post("/products")
+                                            .contentType(MediaType.APPLICATION_JSON)
+                                            .content(JsonUtil.writeValue(TOY_INVALID_MAKER.생성_요청_데이터_생성()))
```
메소드 이름을 한글로 작성하니 읽기 좋군요.

alex-gibeom replied:
테스트 관련 코드는 프러덕션 코드와 다르게, 가독성이 좋은 것이 더 괜찮을 것 같아서 한글로 시도해보았는데 문제 없나보네요!
피드백 감사드립니다 :)

johngrib replied:
저도 테스트 코드에는 한글을 많이 사용하는 편입니다. 한결 읽기 좋죠.

alex-gibeom committed:
test: Product 상품 목록 조회 페이징 테스트 추가

alex-gibeom committed:
refactor: 인덴트 수정 및 HelloControllerTest 삭제

alex-gibeom committed:
feat: User 도메인 Entity 생성

alex-gibeom committed:
test: User 도메인 Entity 단위 테스트 및 Fixture 추가

- Entity Builder 테스트 추가
- Jackson (Equals, HashCode) 테스트 추가
- User 테스트 Fixture 추가

alex-gibeom committed:
feature: User 도메인 JPA PersistenceAdapter 추가

alex-gibeom committed:
test: User 도메인 JPA PersistenceAdapter 단위 테스트 추가

alex-gibeom committed:
feature: User 도메인 FakeObject(인메모리) PersistenceAdapter 추가

alex-gibeom committed:
test: User 도메인 FakeObject(인메모리) PersistenceAdapter 단위 테스트 추가

alex-gibeom committed:
feat: User 도메인 UseCase 추가

alex-gibeom committed:
feat: UserService 생성

alex-gibeom committed:
test: UserService 단위 테스트 - createUser() 기본 성공 테스트 케이스 추가

alex-gibeom committed:
feat: UserService - createUser() 호출 시 회원 저장 후 반환하도록 구현

alex-gibeom committed:
test: UserService 단위 테스트 - createUser() 인자가 null일 경우 실패 테스트 케이스 추가

alex-gibeom committed:
merge: [Day-6] User 도메인 FakeObject를 사용한 Classicist TDD 진행중

johngrib commented on app/src/main/java/com/codesoom/assignment/user/domain/User.java:
```diff
@@ -0,0 +1,32 @@
+@Entity
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+@EqualsAndHashCode
```
`eqauls`와 `hashCode`를 직접 구현해보는 건 좋은 훈련이 됩니다. 롬복 애노테이션이 편리하긴 하지만 구현 방법을 잊지 않도록 종종 이 애노테이션 없이 작업하는 연습도 해보세요.

alex-gibeom replied:
(83b8b43) 권유해주신 대로 primary key인 id를 사용하여 객체 비교 메서드인 `equals`와 `hashCode`를 구현해보았습니다!

johngrib commented on app/src/test/java/com/codesoom/assignment/user/adapter/out/persistence/JpaUserPersistenceAdapterTest.java:
```diff
@@ -0,0 +1,94 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DataJpaTest
+@DisplayName("UserPersistenceAdapter JPA 테스트")
```
JPA 기능을 테스트하는 것이 목적이라면 굳이 작성하지 않아도 됩니다.

alex-gibeom replied:
아하 JPA 기능은 굳이 테스트하지 않아도 되는거였군요. 몰랐네요.
감사합니다 :)

johngrib replied:
프로젝트에 도입을 위한 학습용 테스트라면 괜찮습니다. 하지만 JPA처럼 널리 검증된 라이브러리는 굳이 테스트하지 않는 게 보통이죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/user/adapter/out/persistence/JpaUserPersistenceAdapter.java:
```diff
@@ -1,8 +1,17 @@
+
+    User save(User product);
+
+    void delete(User product);
```
여러 상품들을 삭제하는 관리자용 기능이 필요하다면 어디에서 정의하는 것이 좋을까요?

alex-gibeom replied:
여러 상품들을 삭제하는 관리자용 기능 또한 다른 Product 로직과 동일하게 Product JPA 인터페이스(JpaProductPersistenceAdapter)에 모아놓는 것이 좋을 것 같다는 생각이 듭니다!

이와 별개로 지금 보니 위의 코드에선 User의 JPA 인터페이스인데 매개변수명이 product로 되어있네요..! 한창 1~2주차에 TodoList 개발할 때는 습관적으로 task를 작성하더니, 이젠 또 product를 습관적으로 작성하나봅니다 😂

johngrib replied:
저도 몰랐네요 ㅎㅎㅎ 타입은 User인데 이름이 product라니 ㅎㅎ

johngrib reviewed (approved):
이 정도면 이번주 목표는 달성한 것 같군요.

alex-gibeom committed:
test: JPA 테스트 비활성화

- JPA 기능은 굳이 테스트하지 않아도 됨

alex-gibeom committed:
test: (user-application) 회원 생성 시 회원 목록 카운트 증가 케이스

alex-gibeom committed:
test: (user-application) 회원 정보 수정 성공 케이스

alex-gibeom committed:
test: (user-domain) 기존 회원 정보 수정 성공 케이스

alex-gibeom committed:
feat: (user-application) 회원 정보 수정

alex-gibeom committed:
test: (user-application) 회원 정보 수정 시 회원이 없는 예외 케이스

alex-gibeom committed:
feat: (user-application) 회원 정보 수정 시 회원을 찾을 수 없다면 예외 발생

alex-gibeom committed:
test: (user-application) 회원 삭제 성공 케이스

alex-gibeom committed:
feat: (user-application) 회원 삭제

alex-gibeom committed:
test: (user-application) 회원 삭제 시 회원 목록 카운트 감소 케이스

alex-gibeom committed:
test: (user-application) 회원 삭제 시 회원이 없는 예외 케이스

alex-gibeom committed:
test: (user-presentation) 회원 생성 성공 케이스

alex-gibeom committed:
feat: (user-presentation) 회원 생성 API

alex-gibeom committed:
test: (user-presentation) 회원 수정 성공 케이스

alex-gibeom committed:
feat: (user-presentation) 회원 수정 API

alex-gibeom committed:
test: (user-presentation) 회원 수정 시 회원이 없는 예외 케이스

alex-gibeom committed:
feat: (user-presentation) 회원 수정 시 회원을 찾을 수 없다면 404 코드 반환

alex-gibeom committed:
test: (user-presentation) 회원 삭제 성공 케이스

alex-gibeom committed:
feat: (user-presentation) 회원 삭제 API

alex-gibeom committed:
test: (user-presentation) 회원 삭제 시 회원이 없는 예외 케이스

alex-gibeom committed:
merge: 회원 관리 Classicist TDD로 구현 완료 (확인을 위해  커밋 단위 최소화)

alex-gibeom commented:
오늘 문득 종립 멘토님과 영환 멘토님 덕분에 정말 많이 성장하고 있다는 것이 느껴졌습니다. 감사드립니다 :) 🙇‍♂️🙇‍♂️

(2a39065) 위의 커밋까지 해서 질문이 있습니다.
먼저 커밋 단위를 잘게 쪼갠 이유는, 제가 현재 TDD를 올바른 순서로 하고 있는지 확인 받고 싶었습니다..!

#### [질문 1]
보통 TDD를 구현할 때 Classicist TDD나 Mockist TDD 방식을 진행하는 것으로 알고 있어서 Classicist TDD로 진행해보았습니다.
Classicist TDD는 보통 실제 객체를 통해 테스트를 진행하므로 Inside-Out 방식(Repository -> Service -> Controller)이 좀 더 선호하는 것으로 알고 있는데, 여기서 궁금한건 개발의 단위를 어떤식으로 진행해야 하는지 궁금합니다!

1.  Repository CRUD 다 구현 → Service CRUD 다 구현 → Controller CURD 다 구현
2. 생성하기(Repository → Service → Controller) → 목록 조회(Repository → Service → Controller) → 상세 조회(Repository → Service → Controller)  → …..

1번이 더 괜찮은지, 2번이 더 괜찮은지 아니면 딱히 상관 없는 부분인건지 여쭤보고 싶습니다. (잘 몰라서 이번 커밋에서는 1번으로 진행하였습니다😂)


#### [질문 2]
위의 커밋에서 TDD를 진행하면서 아래와 같은 순서로 진행하였습니다.
- 성공 테스트 케이스 작성 -> 테스트 통과되도록 구현 -> 실패 테스트 케이스 작성 -> 테스트 통과되도록 구현

보통 일반적으로 TDD를 위의 커밋처럼 진행하는건가요? 아니면 조금 다르게 진행하나요?


#### [질문 3]
마지막 질문은 커밋의 단위와, Subject 컨벤션 관련입니다. 회사마다, 팀마다 다르겠지만 경험이 없다보니 궁금해지네요.
보통 현업에서의 커밋 단위를 "기능 단위"로 많이 하는 것으로 알고 있는데 맞게 알고있는건지 궁금합니다. 만약 맞다면 한가지의 기능을 모두 구현한 후 해당 기능 관련된 모든 프러덕션 코드를 한번에 커밋을 하고, 테스트 코드는 또 따로 묶어서 한번에 커밋해서 총 커밋이 2개가 생기는건가요?
또한 커밋 Subject를 제가 오늘 했던 양식이 이상한 방식인지, 괜찮은 방식인지도 궁금합니다!
커밋의 대한 가이드를 부탁드리고 싶습니다.

alex-gibeom commented on app/src/test/java/com/codesoom/assignment/user/adapter/out/persistence/FakeInMemoryUserPersistenceAdapter.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class FakeInMemoryUserPersistenceAdapter implements UserRepository {
+    private static final  AtomicLong id = new AtomicLong(1L);
+    private final List<User> users = new ArrayList<>();
+
+    public User save(User entity) {
+        if (entity == null) {
+            throw new IllegalArgumentException();
+        }
+
+        User user = User.builder()
+                .id(id.getAndIncrement())
+                .name(entity.getName())
+                .email(entity.getEmail())
+                .password(entity.getPassword())
+                .build();
+
+        users.add(user);
+
+        return user;
+    }
+
+    public List<User> findAll() {
+        return users;
+    }
+
+    public Optional<User> findById(Long id) {
+        return users.stream()
+                .filter(user -> user.getId().equals(id))
+                .findFirst();
+    }
+
+    public void delete(User entity) {
+        users.remove(entity);
+    }
+}
```
생각해보니 FakeObject를 사용만 해보고 제대로 여쭤보지 않은 것 같아 질문드리고 싶습니다.
FakeObject를 이런식으로 사용하는 것이 맞는건가요?? 🤔🤔

사용할 때는 아래와 같이 선언하고 있습니다.
```java
class UserServiceTest {
    private UserService userService;
    private FakeInMemoryUserPersistenceAdapter fakeUserRepository;

    @BeforeEach
    void setUpVariable() {
        fakeUserRepository = new FakeInMemoryUserPersistenceAdapter();
        userService = new UserService(fakeUserRepository);
    }
}
```

johngrib replied:
이런 식의 가짜 리포지토리를 만들게 된다면 목적을 분명히 밝히는 것이 좋습니다. 내가 설계한 의도와 전혀 다른 용도로 동료들이 다른 테스트에서 써먹을 수 있기 때문이에요. 지금 저에게 구현 코드를 보여주시는데 아무런 설명이 없어서 저는 `FakeInMemoryUserPersistenceAdapter` 코드를 읽고 목적에 대해 추측을 해야 합니다. 제가 @giibeom 님의 의도를 잘 추측해내면 문제가 없는데 다른 의도라고 생각하게 되면 거기서부터 커뮤니케이션 미스가 발생하게 되겠죠. 그래서 이런 식으로 질문을 할 때에는 상황과 의도를 잘 설명하고, 주석도 충분히 작성하는 것이 좋습니다.

@giibeom 님의 "FakeObject를 이런식으로 사용하는 것이 맞는 건가요?"라는 질문에는 왜 이걸 쓰기로 했고, 왜 이렇게 구현했는지에 대한 설명이 없어요. 그래서 이렇게 사용하는 것이 맞는지 어떤지에 대해 답변을 할 수가 없습니다. 제가 보기에 ArrayList를 갖고 있어서 꽤 작은 수의 사용자 목록을 DB 접속 없이 인메모리로 관리하려는 의도로 만드신 것 같은데, 그 의도가 맞다면 문제 없이 사용할 수 없을 것입니다. 하지만 리포지토리의 구현상 도출되는 몇몇 기능들(예를 들어 auto increment 라던가)을 테스트하고자 만든 것이라면 잘못 만들었다고 할 수 있을 거에요.

alex-gibeom replied:
앗 생각해보니 그러네요. 클래스 상단에 주석을 남겨놓거나, 질문 드릴 때 부연 설명을 했어야됐는데 미처 생각하지 못했던 것 같습니다.

근데 제가 진짜 부족한 질문을 드렸는데, 놀랍게도 의도를 파악해주셨네요..!
`ArrayList를 갖고 있어서 꽤 작은 수의 사용자 목록을 DB 접속 없이 인메모리로 관리하려는 의도`가 맞습니다..!
FakeObject를 썼던 이유는 mock 사용을 지양하기 위해 DB를 대체하는 용도로 사용했습니다. 따라서 Service 테스트 코드는 모두 FakeObject를 통해 DB 대신 사용하고 있습니다.
근데 auto increment 기능 또한 같이 대체하기 위해 AtomicLong을 static으로 선언하여 사용했는데, 잘못 만든걸까요..? 😂

alex-gibeom committed:
test: (user-presentation) 회원 등록 시 유효하지 않은 입력값이 들어온 예외 케이스

- 이름 : 공백
- 이메일 : `@` 없음
- 비밀번호 : 공백

alex-gibeom committed:
feat: (user-presentation) 회원 등록 시 입력값 유효성 검증

- 이름 : null, 공백 허용 안함
- 이메일 : 이메일 형식만 허용
- 비밀번호 : null, 공백 허용 안함

alex-gibeom committed:
test: (user-presentation) 회원 수정 시 유효하지 않은 입력값이 들어온 예외 케이스

- 이름 : 공백
- 이메일 : `@` 없음
- 비밀번호 : 공백

alex-gibeom committed:
feat: (user-presentation) 회원 수정 시 입력값 유효성 검증

- 이름 : null, 공백 허용 안함
- 이메일 : 이메일 형식만 허용
- 비밀번호 : null, 공백 허용 안함

alex-gibeom committed:
refactor: (common-presentation) Exception 관리를 각 도메인 별로 관심사 분리

alex-gibeom committed:
refactor: (common-domain) equals(), hashCode() 직접 구현 및 테스트

alex-gibeom committed:
merge: [Day-7] 회원 Validation, 각 도메인 별 Exception 핸들링 분리

alex-gibeom committed:
fix: (user-persistence) product라고 선언되어 있는 변수명 user로 수정

johngrib commented on app/src/main/java/com/codesoom/assignment/product/domain/Product.java:
```diff
@@ -0,0 +1,81 @@
+            return false;
+        }
+        Product product = (Product) o;
+        return Objects.equals(id, product.id);
```
id만 비교하는 것으로 충분할까요?

alex-gibeom replied:
음.. 사실 처음엔 모든 필드를 다 추가하였다가, "어.. primary key인 id 값만 동일하다면 같은 Entity 데이터라고 볼 수 있지 않을까..?"라고 생각하여서 id만 비교해보았는데 다시 생각해보니 "객체 비교"라는 관점에서는 모든 필드를 추가하는 것이 맞는 것 같습니다 😂


(02e7858) 수정하였습니다. 감사합니다!
```java
@Override
public boolean equals(Object o) {
    if (this == o) {
        return true;
    }
    if (o == null || getClass() != o.getClass()) {
        return false;
    }
    Product product = (Product) o;
    return Objects.equals(id, product.id) && Objects.equals(name, product.name) && Objects.equals(maker, product.maker) && Objects.equals(price, product.price) && Objects.equals(imageUrl, product.imageUrl);
}

@Override
public int hashCode() {
    return Objects.hash(id, name, maker, price, imageUrl);
}
```

johngrib commented:
> 오늘 문득 종립 멘토님과 영환 멘토님 덕분에 정말 많이 성장하고 있다는 것이 느껴졌습니다. 감사드립니다 :) 🙇‍♂️🙇‍♂️
> 
> ([2a39065](https://github.com/CodeSoom/spring-week5-assignment-1/commit/2a3906581090297b2e572e5710ef67a64b56940f)) 위의 커밋까지 해서 질문이 있습니다. 먼저 커밋 단위를 잘게 쪼갠 이유는, 제가 현재 TDD를 올바른 순서로 하고 있는지 확인 받고 싶었습니다..!
> 
> #### [질문 1]
> 보통 TDD를 구현할 때 Classicist TDD나 Mockist TDD 방식을 진행하는 것으로 알고 있어서 Classicist TDD로 진행해보았습니다. Classicist TDD는 보통 실제 객체를 통해 테스트를 진행하므로 Inside-Out 방식(Repository -> Service -> Controller)이 좀 더 선호하는 것으로 알고 있는데, 여기서 궁금한건 개발의 단위를 어떤식으로 진행해야 하는지 궁금합니다!
> 
> 1. Repository CRUD 다 구현 → Service CRUD 다 구현 → Controller CURD 다 구현
> 2. 생성하기(Repository → Service → Controller) → 목록 조회(Repository → Service → Controller) → 상세 조회(Repository → Service → Controller)  → …..
> 
> 1번이 더 괜찮은지, 2번이 더 괜찮은지 아니면 딱히 상관 없는 부분인건지 여쭤보고 싶습니다. (잘 몰라서 이번 커밋에서는 1번으로 진행하였습니다😂)
> 
> #### [질문 2]
> 위의 커밋에서 TDD를 진행하면서 아래와 같은 순서로 진행하였습니다.
> 
> * 성공 테스트 케이스 작성 -> 테스트 통과되도록 구현 -> 실패 테스트 케이스 작성 -> 테스트 통과되도록 구현
> 
> 보통 일반적으로 TDD를 위의 커밋처럼 진행하는건가요? 아니면 조금 다르게 진행하나요?
> 
> #### [질문 3]
> 마지막 질문은 커밋의 단위와, Subject 컨벤션 관련입니다. 회사마다, 팀마다 다르겠지만 경험이 없다보니 궁금해지네요. 보통 현업에서의 커밋 단위를 "기능 단위"로 많이 하는 것으로 알고 있는데 맞게 알고있는건지 궁금합니다. 만약 맞다면 한가지의 기능을 모두 구현한 후 해당 기능 관련된 모든 프러덕션 코드를 한번에 커밋을 하고, 테스트 코드는 또 따로 묶어서 한번에 커밋해서 총 커밋이 2개가 생기는건가요? 또한 커밋 Subject를 제가 오늘 했던 양식이 이상한 방식인지, 괜찮은 방식인지도 궁금합니다! 커밋의 대한 가이드를 부탁드리고 싶습니다.

1, 아무 상관없습니다. TDD 는 R -> S -> C 같은 순서가 전혀 중요하지 않습니다. R/S/C 같은 개념을 아예 써먹지 않는 프로그램을 만들 때에도 TDD를 할 수 있습니다. TDD와 서비스 각 레이어의 개발 순서는 서로 영향을 주지 않습니다.  TDD의 아이디어는 아주 단순해요. 테스트 코드를 항상 먼저 작성하는 것 뿐입니다. 머릿속에 R/S/C 가 너무 강력한 고정관념으로 있어서 TDD 패러다임을 생각할 때도 섞이고 있는 것 같아요. 이거 정말로 아주아주 주의하셔야 합니다. 이런 성향이 심해지면 객체지향과 R/S/C를 구별하지 못하게 되는 경우도 있어요. 만약 컨트롤러를 먼저 만들기로 결심했는데 TDD를 하고자 한다면 컨트롤러의 테스트를 먼저 만들면 됩니다. 프로젝트 전체를 머릿속에서 다 만들고 TDD를 하려고 하면 잘 안 될 거에요. 오히려 TDD 방식이 거슬릴 수도 있습니다. TDD를 하고 싶다면 개발하고 싶은 순서대로 개발하면 됩니다. 대신 각각의 컴포넌트를 코딩하기 직전에 테스트를 먼저 작성해 보세요.

2, 아무 상관없습니다. 일반적으로 TDD를 어떻게 하는지는 아무도 모릅니다. 사람마다 다 다르게 해요. TDD는 그냥 테스트 코드를 먼저 작성해서 테스트가 실패할 수 밖에 없는 상황을 만들고, 테스트코드의 피드백을 받으면서 코딩하고, 성공하면 또 테스트코드를 추가하는 걸 반복하는 방법론입니다. 그래서 성공 케이스를 먼저 작성하는지 실패 케이스를 먼저 작성하는지는 하나도 중요하지 않아요. 잘 모르겠으면 그냥 제일 먼저 생각나는 테스트 케이스를 테스트 코드로 만들면서 작업하면 됩니다. 

3, 이런 답변을 드리는 것이 굉장히 안타까운데요, "보통 현업에서는..." 같은 질문을 하면 좋은 대답을 드리기 어렵습니다. 왜냐하면 회사마다 다 다르게 하거든요. 그리고 상당수의 회사가 얼렁뚱땅 만든 프로세스로 대충 굴러가는 경우가 엄청 많습니다. 그래서 "보통 현업에서는" 처럼 질문을 하시면 "그런 규칙 없이 사람마다 다 다르게 합니다"라고 대답하게 되죠. 하지만 이런 답변을 기대하신 게 아닐 거에요. 아마 베스트 프랙티스가 궁금하셔서 질문하셨을 거고, 그걸 답변으로 들으면 제가 이야기한 베스트 프랙티스를 반복 연습해서 익숙해지려고 하신 질문이겠죠! 그런데 이거 굉장히 위험한 접근입니다. 만약 제가 "이게 베스트 프랙티스입니다"라고 말씀을 드려서 그걸 엄청나게 연습하셨다고 합시다. 그런데 열심히 연습했는데 3달 만에 뭔가 취약점이 발견되었거나 유행이 바뀌었다던가 해서 베스트 프랙티스가 바뀌면 어떻게 할까요? 연습한 것이 아깝게 되지 않을까요?

우리 개발자들의 딜레마가 여기에 있습니다. 베스트 프랙티스가 없다시피 합니다. 있다 해도 보통은 개발팀에서 스스로 결정하고, 뒤에 입사하는 사람들은 그걸 따르게 되죠. 그래서 베스트 프랙티스가 회사마다 달라요. 어떤 회사는 커밋이 많아도 신경쓰지 않고, 어떤 회사는 기능별로 잘 분리한 커밋 목록을 선호하고, 어떤 회사는 머지만 사용하고, 어떤 회사는 강력하게 리베이스를 사용하도록 강제합니다. 그래서 이 질문은 답변하기가 애매합니다. 베스트 프랙티스는 이것입니다 라고 알려드렸는데, 그 방식을 싫어하는 회사에 다니시게 되면 제 조언이 적절하지 않은 게 되겠지요! 

차라리 제 취향을 물어보신다면 답변이 쉽습니다. 

제 취향은 이렇습니다.

- PR 하나는 작업에 대한 보고서이자 설계이자 제안서입니다. 
- PR에 포함된 commit 하나하나는 해당 제안서에서 이야기하고자 하는 작업들을 이해하기 쉽게 쪼개놓은 것입니다.
- 따라서 할 일 목록을 먼저 종이에 써보세요. 이해하기 쉽게 써야 합니다. 너무 많아도 안 되고 너무 적어도 안 됩니다.
- 그 할 일 목록의 한 줄 한 줄이 커밋 메시지가 되도록 작업을 합니다.

그리고 커밋이 너무 많으면.... 리뷰하기 힘듭니다 ㅎㅎ

johngrib commented on app/src/main/java/com/codesoom/assignment/user/adapter/in/web/dto/request/UserCreateRequestDto.java:
```diff
@@ -5,10 +5,22 @@
+
+    // Regular Expression by RFC 5322 for Email Validation
+    // Ref: https://www.baeldung.com/java-email-validation-regex#regular-expression-by-rfc-5322-for-email-validation
+    @Pattern(regexp = "^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$",
+            message = "올바른 이메일 형식으로 입력하세요")
```
`@Email` 을 사용할 수 있는지에 대해서도 조사해 보세요. 이런 식으로 정규식을 사용할 땐 기차 레일 다이어그램을 그리거나, 이 정규식을 검증하는 테스트 코드를 함께 PR에 포함시켜 주세요. 그렇게 하지 않으면 리뷰어가(동료가) 힘들어하거나 대충 보게 됩니다.

alex-gibeom replied:
아하 java validation에 `@Email`이 있네요! 이메일 형식을 검증해야겠다라는 생각이 자연스럽게 `@Patter`을 찾게 됐네요.

`@Email`이 기본적으로 null을 유효하다고 판단하니, `@NotBlank`를 같이 사용해주어야 겠네요.
<img src="https://user-images.githubusercontent.com/59248326/201535327-71aca52d-72bf-4e3f-96a0-2c92fac44f6a.png" width="70%" />

<br>

(5d97354) 수정하였습니다. 감사합니다!
```java
@NotBlank(message = "이메일은 필수 입력 값입니다.")
@Email(message = "이메일 형식에 맞지 않습니다.")
private String email;
```


추가로 패턴을 사용할 때는 기차 레일 다이어그램을 꼭 사용하겠습니다!
검색하니 종립 멘토님께서 트위터에서 추천하신 [사이트](https://regexper.com/)가 바로 검색되네요 ㅎㅎ
사이트 북마크 해놔야겠습니다 :)

<img src="https://user-images.githubusercontent.com/59248326/201535508-dc8e7efc-6893-447b-8018-1b64bb0d1b8c.png" width="70%" />

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해 주세요.

alex-gibeom committed:
fix: (user-domain) equals(), hashCode() id로만 비교한 로직을 모든 필드로 비교하도록 변경

alex-gibeom committed:
fix: (user-domain) 이메일 검증 `@Pattern`에서 `@Email`로 변경
